compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -605,70 +609,82 @@ private void defineTypeNodes(List<? extends TypeDefinition> typeDefs, SymbolEnv
                 }
                 break;
             case CONSTRAINED_TYPE:
-                checkErrors(unresolvedType, ((BLangConstrainedType) currentTypeNode).constraint, visitedNodes);
+                checkErrors(unresolvedType, ((BLangConstrainedType) currentTypeNode).constraint, visitedNodes, true);
+                break;
+            case TABLE_TYPE:
+                checkErrors(unresolvedType, ((BLangTableTypeNode) currentTypeNode).constraint, visitedNodes, true);
                 break;
             case USER_DEFINED_TYPE:
                 String currentTypeNodeName = ((BLangUserDefinedType) currentTypeNode).typeName.value;
                 // Skip all types defined as anonymous types.
                 if (currentTypeNodeName.startsWith("$")) {
                     return;
                 }
-
-                if (unresolvedTypeNodeName.equals(currentTypeNodeName)) {
-                    // Cyclic dependency detected. We need to add the `unresolvedTypeNodeName` or the
-                    // `memberTypeNodeName` to the end of the list to complete the cyclic dependency when
-                    // printing the error.
-                    visitedNodes.push(currentTypeNodeName);
-                    dlog.error((DiagnosticPos) unresolvedType.getPosition(), DiagnosticCode.CYCLIC_TYPE_REFERENCE,
-                            visitedNodes);
-                    // We need to remove the last occurrence since we use this list in a recursive call.
-                    // Otherwise, unwanted types will get printed in the cyclic dependency error.
-                    visitedNodes.pop();
-                } else if (visitedNodes.contains(currentTypeNodeName)) {
-                    // Cyclic dependency detected. But in here, all the types in the list might not be necessary for the
-                    // cyclic dependency error message.
-                    //
-                    // Eg - A -> B -> C -> B // Last B is what we are currently checking
-                    //
-                    // In such case, we create a new list with relevant type names.
-                    int i = visitedNodes.indexOf(currentTypeNodeName);
-                    List<String> dependencyList = new ArrayList<>(visitedNodes.size() - i);
-                    for (; i < visitedNodes.size(); i++) {
-                        dependencyList.add(visitedNodes.get(i));
+                boolean sameTypeNode = unresolvedTypeNodeName.equals(currentTypeNodeName);
+                boolean isVisitedNode = visitedNodes.contains(currentTypeNodeName);
+                boolean typeDef =  unresolvedType.getKind() == NodeKind.TYPE_DEFINITION;
+
+                if (typeDef && (sameTypeNode || isVisitedNode)) {
+                    BLangTypeDefinition typeDefinition = (BLangTypeDefinition) unresolvedType;
+                    if (fromStructuredType && typeDefinition.getTypeNode().getKind() == NodeKind.UNION_TYPE_NODE) {
```

Comment:
Why do we need this `fromStructuredType` check?

Reply:
To avoid scenarios, such as the following
```ballerina
type A int|A;
```
and to only allow scenarios where the cyclic is coming from a structural type.
```ballerina
type A int|map<A>
```